File: src/algorithms.py
"""
Pathfinding Algorithms Module
Implements DFS, BFS, Dijkstra, A*, and Greedy Best-First Search
"""
from typing import List, Tuple, Optional, Callable, Set, Dict
from collections import deque
import heapq
import math
from src.maze import Maze, CellType
from src.metrics import MetricsTracker, AlgorithmMetrics


class PathfindingAlgorithms:
    """
    Collection of pathfinding algorithms with visualization support
    """
    
    def __init__(self, maze: Maze):
        """
        Initialize pathfinding algorithms with a maze
        
        Args:
            maze: Maze object to solve
        """
        self.maze = maze
        self.visualization_callback: Optional[Callable] = None
    
    def set_visualization_callback(self, callback: Callable):
        """
        Set callback function for visualization
        
        Args:
            callback: Function to call for each step (row, col, state)
        """
        self.visualization_callback = callback
    
    def _visualize_step(self, row: int, col: int, state: str):
        """Call visualization callback if set"""
        if self.visualization_callback:
            self.visualization_callback(row, col, state)
    
    def _reconstruct_path(self, came_from: Dict[Tuple[int, int], Tuple[int, int]], 
                         current: Tuple[int, int]) -> List[Tuple[int, int]]:
        """
        Reconstruct path from start to end using came_from dictionary
        
        Args:
            came_from: Dictionary mapping each node to its predecessor
            current: End node
            
        Returns:
            List of coordinates representing the path
        """
        path = [current]
        while current in came_from:
            current = came_from[current]
            path.append(current)
        path.reverse()
        return path
    
    # ===== DFS (Stack-based) =====
    def dfs(self) -> AlgorithmMetrics:
        """
        Depth-First Search using explicit stack
        
        Time Complexity: O(V + E) where V = vertices, E = edges
        Space Complexity: O(V) for the stack and visited set
        
        Returns:
            AlgorithmMetrics object with performance data
        """
        tracker = MetricsTracker()
        tracker.start_tracking()
        
        start = self.maze.start
        end = self.maze.end
        
        # Stack for DFS (LIFO)
        stack = [start]
        visited: Set[Tuple[int, int]] = {start}
        came_from: Dict[Tuple[int, int], Tuple[int, int]] = {}
        
        found = False
        
        while stack:
            tracker.update_frontier_size(len(stack))
            current = stack.pop()
            tracker.increment_nodes()
            
            self._visualize_step(current[0], current[1], 'exploring')
            
            if current == end:
                found = True
                break
            
            # Explore neighbors
            for neighbor in self.maze.get_neighbors(current[0], current[1]):
                if neighbor not in visited:
                    visited.add(neighbor)
                    stack.append(neighbor)
                    came_from[neighbor] = current
                    self._visualize_step(neighbor[0], neighbor[1], 'visited')
        
        # Reconstruct path
        path = []
        if found:
            path = self._reconstruct_path(came_from, end)
            for row, col in path:
                if (row, col) != start and (row, col) != end:
                    self._visualize_step(row, col, 'path')
        
        return tracker.create_metrics(
            algorithm_name="DFS (Stack)",
            path=path,
            time_complexity="O(V + E)",
            space_complexity="O(V)",
            is_optimal=False  # DFS does not guarantee shortest path
        )
    
    # ===== BFS (Queue-based) =====
    def bfs(self) -> AlgorithmMetrics:
        """
        Breadth-First Search using queue
        
        Time Complexity: O(V + E)
        Space Complexity: O(V)
        
        Returns:
            AlgorithmMetrics object with performance data
        """
        tracker = MetricsTracker()
        tracker.start_tracking()
        
        start = self.maze.start
        end = self.maze.end
        
        # Queue for BFS (FIFO)
        queue = deque([start])
        visited: Set[Tuple[int, int]] = {start}
        came_from: Dict[Tuple[int, int], Tuple[int, int]] = {}
        
        found = False
        
        while queue:
            tracker.update_frontier_size(len(queue))
            current = queue.popleft()
            tracker.increment_nodes()
            
            self._visualize_step(current[0], current[1], 'exploring')
            
            if current == end:
                found = True
                break
            
            # Explore neighbors
            for neighbor in self.maze.get_neighbors(current[0], current[1]):
                if neighbor not in visited:
                    visited.add(neighbor)
                    queue.append(neighbor)
                    came_from[neighbor] = current
                    self._visualize_step(neighbor[0], neighbor[1], 'visited')
        
        # Reconstruct path
        path = []
        if found:
            path = self._reconstruct_path(came_from, end)
            for row, col in path:
                if (row, col) != start and (row, col) != end:
                    self._visualize_step(row, col, 'path')
        
        return tracker.create_metrics(
            algorithm_name="BFS (Queue)",
            path=path,
            time_complexity="O(V + E)",
            space_complexity="O(V)",
            is_optimal=True  # BFS guarantees shortest path in unweighted graphs
        )
    
    # ===== Dijkstra's Algorithm (Priority Queue / Min-Heap) =====
    def dijkstra(self) -> AlgorithmMetrics:
        """
        Dijkstra's Algorithm using Min-Heap (Priority Queue)
        
        Time Complexity: O((V + E) log V)
        Space Complexity: O(V)
        
        Returns:
            AlgorithmMetrics object with performance data
        """
        tracker = MetricsTracker()
        tracker.start_tracking()
        
        start = self.maze.start
        end = self.maze.end
        
        # Priority queue: (cost, node)
        pq = [(0, start)]
        visited: Set[Tuple[int, int]] = set()
        came_from: Dict[Tuple[int, int], Tuple[int, int]] = {}
        cost_so_far: Dict[Tuple[int, int], float] = {start: 0}
        
        found = False
        
        while pq:
            tracker.update_frontier_size(len(pq))
            current_cost, current = heapq.heappop(pq)
            
            if current in visited:
                continue
            
            visited.add(current)
            tracker.increment_nodes()
            
            self._visualize_step(current[0], current[1], 'exploring')
            
            if current == end:
                found = True
                break
            
            # Explore neighbors
            for neighbor in self.maze.get_neighbors(current[0], current[1]):
                new_cost = current_cost + self.maze.get_edge_weight(current, neighbor)
                
                if neighbor not in cost_so_far or new_cost < cost_so_far[neighbor]:
                    cost_so_far[neighbor] = new_cost
                    heapq.heappush(pq, (new_cost, neighbor))
                    came_from[neighbor] = current
                    self._visualize_step(neighbor[0], neighbor[1], 'visited')
        
        # Reconstruct path
        path = []
        if found:
            path = self._reconstruct_path(came_from, end)
            for row, col in path:
                if (row, col) != start and (row, col) != end:
                    self._visualize_step(row, col, 'path')
        
        return tracker.create_metrics(
            algorithm_name="Dijkstra (Min-Heap)",
            path=path,
            time_complexity="O((V + E) log V)",
            space_complexity="O(V)",
            is_optimal=True  # Dijkstra guarantees optimal path
        )
    
    # ===== Heuristic Functions =====
    @staticmethod
    def manhattan_distance(pos1: Tuple[int, int], pos2: Tuple[int, int]) -> float:
        """
        Manhattan distance heuristic (L1 norm)
        Good for 4-directional movement
        """
        return abs(pos1[0] - pos2[0]) + abs(pos1[1] - pos2[1])
    
    @staticmethod
    def euclidean_distance(pos1: Tuple[int, int], pos2: Tuple[int, int]) -> float:
        """
        Euclidean distance heuristic (L2 norm)
        Straight-line distance
        """
        return math.sqrt((pos1[0] - pos2[0])**2 + (pos1[1] - pos2[1])**2)
    
    @staticmethod
    def chebyshev_distance(pos1: Tuple[int, int], pos2: Tuple[int, int]) -> float:
        """
        Chebyshev distance heuristic (L∞ norm)
        Good for 8-directional movement
        """
        return max(abs(pos1[0] - pos2[0]), abs(pos1[1] - pos2[1]))
    
    # ===== A* Algorithm =====
    def astar(self, heuristic: str = 'manhattan') -> AlgorithmMetrics:
        """
        A* Algorithm with selectable heuristic
        
        Time Complexity: O(E) in worst case, O(b^d) average
        Space Complexity: O(V)
        
        Args:
            heuristic: Heuristic function ('manhattan', 'euclidean', 'chebyshev')
            
        Returns:
            AlgorithmMetrics object with performance data
        """
        tracker = MetricsTracker()
        tracker.start_tracking()
        
        # Select heuristic function
        heuristic_func = {
            'manhattan': self.manhattan_distance,
            'euclidean': self.euclidean_distance,
            'chebyshev': self.chebyshev_distance
        }.get(heuristic, self.manhattan_distance)
        
        start = self.maze.start
        end = self.maze.end
        
        # Priority queue: (f_score, g_score, node)
        pq = [(heuristic_func(start, end), 0, start)]
        visited: Set[Tuple[int, int]] = set()
        came_from: Dict[Tuple[int, int], Tuple[int, int]] = {}
        g_score: Dict[Tuple[int, int], float] = {start: 0}
        
        found = False
        
        while pq:
            tracker.update_frontier_size(len(pq))
            f, current_g, current = heapq.heappop(pq)
            
            if current in visited:
                continue
            
            visited.add(current)
            tracker.increment_nodes()
            
            self._visualize_step(current[0], current[1], 'exploring')
            
            if current == end:
                found = True
                break
            
            # Explore neighbors
            for neighbor in self.maze.get_neighbors(current[0], current[1]):
                new_g = current_g + self.maze.get_edge_weight(current, neighbor)
                
                if neighbor not in g_score or new_g < g_score[neighbor]:
                    g_score[neighbor] = new_g
                    h = heuristic_func(neighbor, end)
                    f_score = new_g + h
                    heapq.heappush(pq, (f_score, new_g, neighbor))
                    came_from[neighbor] = current
                    self._visualize_step(neighbor[0], neighbor[1], 'visited')
        
        # Reconstruct path
        path = []
        if found:
            path = self._reconstruct_path(came_from, end)
            for row, col in path:
                if (row, col) != start and (row, col) != end:
                    self._visualize_step(row, col, 'path')
        
        return tracker.create_metrics(
            algorithm_name=f"A* ({heuristic.capitalize()})",
            path=path,
            time_complexity="O(b^d)",
            space_complexity="O(V)",
            is_optimal=True,  # A* is optimal with admissible heuristic
            heuristic=heuristic.capitalize()
        )
    
    # ===== Greedy Best-First Search =====
    def greedy_best_first(self, heuristic: str = 'manhattan') -> AlgorithmMetrics:
        """
        Greedy Best-First Search
        Uses only heuristic (no path cost consideration)
        
        Time Complexity: O(b^m) where m is maximum depth
        Space Complexity: O(V)
        
        Args:
            heuristic: Heuristic function ('manhattan', 'euclidean', 'chebyshev')
            
        Returns:
            AlgorithmMetrics object with performance data
        """
        tracker = MetricsTracker()
        tracker.start_tracking()
        
        # Select heuristic function
        heuristic_func = {
            'manhattan': self.manhattan_distance,
            'euclidean': self.euclidean_distance,
            'chebyshev': self.chebyshev_distance
        }.get(heuristic, self.manhattan_distance)
        
        start = self.maze.start
        end = self.maze.end
        
        # Priority queue: (h_score, node)
        pq = [(heuristic_func(start, end), start)]
        visited: Set[Tuple[int, int]] = set()
        came_from: Dict[Tuple[int, int], Tuple[int, int]] = {}
        
        found = False
        
        while pq:
            tracker.update_frontier_size(len(pq))
            h, current = heapq.heappop(pq)
            
            if current in visited:
                continue
            
            visited.add(current)
            tracker.increment_nodes()
            
            self._visualize_step(current[0], current[1], 'exploring')
            
            if current == end:
                found = True
                break
            
            # Explore neighbors
            for neighbor in self.maze.get_neighbors(current[0], current[1]):
                if neighbor not in visited:
                    h_score = heuristic_func(neighbor, end)
                    heapq.heappush(pq, (h_score, neighbor))
                    if neighbor not in came_from:
                        came_from[neighbor] = current
                    self._visualize_step(neighbor[0], neighbor[1], 'visited')
        
        # Reconstruct path
        path = []
        if found:
            path = self._reconstruct_path(came_from, end)
            for row, col in path:
                if (row, col) != start and (row, col) != end:
                    self._visualize_step(row, col, 'path')
        
        return tracker.create_metrics(
            algorithm_name=f"Greedy Best-First ({heuristic.capitalize()})",
            path=path,
            time_complexity="O(b^m)",
            space_complexity="O(V)",
            is_optimal=False,  # Greedy is not optimal
            heuristic=heuristic.capitalize()
        )
    
    # ===== Bidirectional Search (Bonus) =====
    def bidirectional_search(self) -> AlgorithmMetrics:
        """
        Bidirectional BFS - searches from both start and end simultaneously
        
        Time Complexity: O(b^(d/2))
        Space Complexity: O(b^(d/2))
        
        Returns:
            AlgorithmMetrics object with performance data
        """
        tracker = MetricsTracker()
        tracker.start_tracking()
        
        start = self.maze.start
        end = self.maze.end
        
        # Two queues for forward and backward search
        forward_queue = deque([start])
        backward_queue = deque([end])
        
        forward_visited: Dict[Tuple[int, int], Tuple[int, int]] = {start: None}
        backward_visited: Dict[Tuple[int, int], Tuple[int, int]] = {end: None}
        
        meeting_point = None
        
        while forward_queue and backward_queue:
            tracker.update_frontier_size(len(forward_queue) + len(backward_queue))
            
            # Forward search
            if forward_queue:
                current = forward_queue.popleft()
                tracker.increment_nodes()
                self._visualize_step(current[0], current[1], 'exploring')
                
                # Check if paths meet
                if current in backward_visited:
                    meeting_point = current
                    break
                
                for neighbor in self.maze.get_neighbors(current[0], current[1]):
                    if neighbor not in forward_visited:
                        forward_visited[neighbor] = current
                        forward_queue.append(neighbor)
                        self._visualize_step(neighbor[0], neighbor[1], 'visited')
            
            # Backward search
            if backward_queue:
                current = backward_queue.popleft()
                tracker.increment_nodes()
                self._visualize_step(current[0], current[1], 'exploring')
                
                # Check if paths meet
                if current in forward_visited:
                    meeting_point = current
                    break
                
                for neighbor in self.maze.get_neighbors(current[0], current[1]):
                    if neighbor not in backward_visited:
                        backward_visited[neighbor] = current
                        backward_queue.append(neighbor)
                        self._visualize_step(neighbor[0], neighbor[1], 'visited')
        
        # Reconstruct path
        path = []
        if meeting_point:
            # Build forward path
            forward_path = []
            current = meeting_point
            while current is not None:
                forward_path.append(current)
                current = forward_visited[current]
            forward_path.reverse()
            
            # Build backward path
            backward_path = []
            current = backward_visited[meeting_point]
            while current is not None:
                backward_path.append(current)
                current = backward_visited[current]
            
            path = forward_path + backward_path
            
            for row, col in path:
                if (row, col) != start and (row, col) != end:
                    self._visualize_step(row, col, 'path')
        
        return tracker.create_metrics(
            algorithm_name="Bidirectional BFS",
            path=path,
            time_complexity="O(b^(d/2))",
            space_complexity="O(b^(d/2))",
            is_optimal=True
        )

File: algorithms.js
class PathfindingAlgorithms {
    static manhattanDistance(a, b) {
        return Math.abs(a.row - b.row) + Math.abs(a.col - b.col);
    }

    static euclideanDistance(a, b) {
        return Math.sqrt(Math.pow(a.row - b.row, 2) + Math.pow(a.col - b.col, 2));
    }

    static async dfs(maze, onVisit) {
        const start = performance.now();
        const stack = [maze.start];
        const visited = new Set();
        const parent = new Map();
        let nodesExplored = 0;

        visited.add(`${maze.start.row},${maze.start.col}`);

        while (stack.length > 0) {
            const current = stack.pop();
            nodesExplored++;

            await onVisit(current, 'exploring');

            if (current.row === maze.end.row && current.col === maze.end.col) {
                const path = this.reconstructPath(parent, maze.start, maze.end);
                const time = performance.now() - start;
                return { path, nodesExplored, time, optimal: false };
            }

            const neighbors = maze.getNeighbors(current.row, current.col);
            for (const neighbor of neighbors) {
                const key = `${neighbor.row},${neighbor.col}`;
                if (!visited.has(key)) {
                    visited.add(key);
                    parent.set(key, current);
                    stack.push(neighbor);
                }
            }
        }

        const time = performance.now() - start;
        return { path: null, nodesExplored, time, optimal: false };
    }

    static async bfs(maze, onVisit) {
        const start = performance.now();
        const queue = [maze.start];
        const visited = new Set();
        const parent = new Map();
        let nodesExplored = 0;

        visited.add(`${maze.start.row},${maze.start.col}`);

        while (queue.length > 0) {
            const current = queue.shift();
            nodesExplored++;

            await onVisit(current, 'exploring');

            if (current.row === maze.end.row && current.col === maze.end.col) {
                const path = this.reconstructPath(parent, maze.start, maze.end);
                const time = performance.now() - start;
                return { path, nodesExplored, time, optimal: true };
            }

            const neighbors = maze.getNeighbors(current.row, current.col);
            for (const neighbor of neighbors) {
                const key = `${neighbor.row},${neighbor.col}`;
                if (!visited.has(key)) {
                    visited.add(key);
                    parent.set(key, current);
                    queue.push(neighbor);
                }
            }
        }

        const time = performance.now() - start;
        return { path: null, nodesExplored, time, optimal: true };
    }

    static async dijkstra(maze, onVisit) {
        const start = performance.now();
        const distances = new Map();
        const parent = new Map();
        const visited = new Set();
        const pq = new PriorityQueue();
        let nodesExplored = 0;

        const startKey = `${maze.start.row},${maze.start.col}`;
        distances.set(startKey, 0);
        pq.enqueue(maze.start, 0);

        while (!pq.isEmpty()) {
            const current = pq.dequeue();
            const currentKey = `${current.row},${current.col}`;

            if (visited.has(currentKey)) continue;
            visited.add(currentKey);
            nodesExplored++;

            await onVisit(current, 'exploring');

            if (current.row === maze.end.row && current.col === maze.end.col) {
                const path = this.reconstructPath(parent, maze.start, maze.end);
                const time = performance.now() - start;
                return { path, nodesExplored, time, optimal: true };
            }

            const neighbors = maze.getNeighbors(current.row, current.col);
            for (const neighbor of neighbors) {
                const neighborKey = `${neighbor.row},${neighbor.col}`;
                const newDist = distances.get(currentKey) + 1;

                if (!distances.has(neighborKey) || newDist < distances.get(neighborKey)) {
                    distances.set(neighborKey, newDist);
                    parent.set(neighborKey, current);
                    pq.enqueue(neighbor, newDist);
                }
            }
        }

        const time = performance.now() - start;
        return { path: null, nodesExplored, time, optimal: true };
    }

    static async astar(maze, onVisit, heuristic = 'manhattan') {
        const start = performance.now();
        const gScore = new Map();
        const fScore = new Map();
        const parent = new Map();
        const visited = new Set();
        const pq = new PriorityQueue();
        let nodesExplored = 0;

        const heuristicFunc = heuristic === 'euclidean' ? 
            this.euclideanDistance : this.manhattanDistance;

        const startKey = `${maze.start.row},${maze.start.col}`;
        gScore.set(startKey, 0);
        fScore.set(startKey, heuristicFunc(maze.start, maze.end));
        pq.enqueue(maze.start, fScore.get(startKey));

        while (!pq.isEmpty()) {
            const current = pq.dequeue();
            const currentKey = `${current.row},${current.col}`;

            if (visited.has(currentKey)) continue;
            visited.add(currentKey);
            nodesExplored++;

            await onVisit(current, 'exploring');

            if (current.row === maze.end.row && current.col === maze.end.col) {
                const path = this.reconstructPath(parent, maze.start, maze.end);
                const time = performance.now() - start;
                return { path, nodesExplored, time, optimal: true, heuristic };
            }

            const neighbors = maze.getNeighbors(current.row, current.col);
            for (const neighbor of neighbors) {
                const neighborKey = `${neighbor.row},${neighbor.col}`;
                const tentativeGScore = gScore.get(currentKey) + 1;

                if (!gScore.has(neighborKey) || tentativeGScore < gScore.get(neighborKey)) {
                    parent.set(neighborKey, current);
                    gScore.set(neighborKey, tentativeGScore);
                    const h = heuristicFunc(neighbor, maze.end);
                    fScore.set(neighborKey, tentativeGScore + h);
                    pq.enqueue(neighbor, fScore.get(neighborKey));
                }
            }
        }

        const time = performance.now() - start;
        return { path: null, nodesExplored, time, optimal: true, heuristic };
    }

    static async greedy(maze, onVisit) {
        const start = performance.now();
        const parent = new Map();
        const visited = new Set();
        const pq = new PriorityQueue();
        let nodesExplored = 0;

        pq.enqueue(maze.start, this.manhattanDistance(maze.start, maze.end));
        visited.add(`${maze.start.row},${maze.start.col}`);

        while (!pq.isEmpty()) {
            const current = pq.dequeue();
            nodesExplored++;

            await onVisit(current, 'exploring');

            if (current.row === maze.end.row && current.col === maze.end.col) {
                const path = this.reconstructPath(parent, maze.start, maze.end);
                const time = performance.now() - start;
                return { path, nodesExplored, time, optimal: false };
            }

            const neighbors = maze.getNeighbors(current.row, current.col);
            for (const neighbor of neighbors) {
                const key = `${neighbor.row},${neighbor.col}`;
                if (!visited.has(key)) {
                    visited.add(key);
                    parent.set(key, current);
                    const h = this.manhattanDistance(neighbor, maze.end);
                    pq.enqueue(neighbor, h);
                }
            }
        }

        const time = performance.now() - start;
        return { path: null, nodesExplored, time, optimal: false };
    }

    static async bidirectional(maze, onVisit) {
        const start = performance.now();
        const queueStart = [maze.start];
        const queueEnd = [maze.end];
        const visitedStart = new Set([`${maze.start.row},${maze.start.col}`]);
        const visitedEnd = new Set([`${maze.end.row},${maze.end.col}`]);
        const parentStart = new Map();
        const parentEnd = new Map();
        let nodesExplored = 0;

        while (queueStart.length > 0 && queueEnd.length > 0) {
            // Expand from start
            if (queueStart.length > 0) {
                const current = queueStart.shift();
                nodesExplored++;
                await onVisit(current, 'exploring');

                const currentKey = `${current.row},${current.col}`;
                if (visitedEnd.has(currentKey)) {
                    const path = this.reconstructBidirectionalPath(
                        parentStart, parentEnd, maze.start, current, maze.end
                    );
                    const time = performance.now() - start;
                    return { path, nodesExplored, time, optimal: true };
                }

                const neighbors = maze.getNeighbors(current.row, current.col);
                for (const neighbor of neighbors) {
                    const key = `${neighbor.row},${neighbor.col}`;
                    if (!visitedStart.has(key)) {
                        visitedStart.add(key);
                        parentStart.set(key, current);
                        queueStart.push(neighbor);
                    }
                }
            }

            // Expand from end
            if (queueEnd.length > 0) {
                const current = queueEnd.shift();
                nodesExplored++;
                await onVisit(current, 'exploring');

                const currentKey = `${current.row},${current.col}`;
                if (visitedStart.has(currentKey)) {
                    const path = this.reconstructBidirectionalPath(
                        parentStart, parentEnd, maze.start, current, maze.end
                    );
                    const time = performance.now() - start;
                    return { path, nodesExplored, time, optimal: true };
                }

                const neighbors = maze.getNeighbors(current.row, current.col);
                for (const neighbor of neighbors) {
                    const key = `${neighbor.row},${neighbor.col}`;
                    if (!visitedEnd.has(key)) {
                        visitedEnd.add(key);
                        parentEnd.set(key, current);
                        queueEnd.push(neighbor);
                    }
                }
            }
        }

        const time = performance.now() - start;
        return { path: null, nodesExplored, time, optimal: true };
    }

    static reconstructPath(parent, start, end) {
        const path = [];
        let current = end;
        const startKey = `${start.row},${start.col}`;

        while (current) {
            path.unshift(current);
            const key = `${current.row},${current.col}`;
            if (key === startKey) break;
            current = parent.get(key);
        }

        return path;
    }

    static reconstructBidirectionalPath(parentStart, parentEnd, start, meeting, end) {
        const pathStart = [];
        let current = meeting;
        const startKey = `${start.row},${start.col}`;

        while (current) {
            pathStart.unshift(current);
            const key = `${current.row},${current.col}`;
            if (key === startKey) break;
            current = parentStart.get(key);
        }

        const pathEnd = [];
        current = parentEnd.get(`${meeting.row},${meeting.col}`);
        while (current) {
            pathEnd.push(current);
            const key = `${current.row},${current.col}`;
            const endKey = `${end.row},${end.col}`;
            if (key === endKey) break;
            current = parentEnd.get(key);
        }

        return [...pathStart, ...pathEnd];
    }
}

class PriorityQueue {
    constructor() {
        this.values = [];
    }

    enqueue(item, priority) {
        this.values.push({ item, priority });
        this.values.sort((a, b) => a.priority - b.priority);
    }

    dequeue() {
        return this.values.shift()?.item;
    }

    isEmpty() {
        return this.values.length === 0;
    }
}
